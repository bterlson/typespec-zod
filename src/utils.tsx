import { Children } from "@alloy-js/core/jsx-runtime";
import { FunctionCallExpression } from "@alloy-js/typescript";
import { Type } from "@typespec/compiler";
import { $ } from "@typespec/compiler/experimental/typekit";

export const refkeySym = Symbol.for("typespec-zod.refkey");

/**
 * Returns true if the given type is a declaration or an instantiation of a
 * declaration.
 */
export function isDeclaration(type: Type): boolean {
  switch (type.kind) {
    case "Namespace":
    case "Interface":
    case "Operation":
    case "EnumMember":
      // TODO: this should reference the enum member via
      // target.enum.Name
      return false;
    case "UnionVariant":
      return false;

    case "Model":
      if (($.array.is(type) || $.record.is(type)) && isBuiltIn(type)) {
        return false;
      }

      return Boolean(type.name);
    case "Union":
      return Boolean(type.name);
    case "Enum":
      return true;
    case "Scalar":
      return true;
    default:
      return false;
  }
}

// typekit doesn't consider things which have properties as records
// even though they are?
export function isRecord(type: Type): boolean {
  return (
    type.kind === "Model" &&
    !!type.indexer &&
    type.indexer.key === $.builtin.string
  );
}

export function shouldReference(type: Type) {
  return isDeclaration(type) && !isBuiltIn(type);
}

export function isBuiltIn(type: Type) {
  if (!("namespace" in type) || type.namespace === undefined) {
    return false;
  }

  const globalNs = $.program.getGlobalNamespaceType();
  let tln = type.namespace;
  if (tln === globalNs) {
    return false;
  }

  while (tln.namespace !== globalNs) {
    tln = tln.namespace!;
  }

  return tln === globalNs.namespaces.get("TypeSpec");
}

/**
 * This API takes an array of types and returns those types in an array of
 * cyclesets. A cycleset is a group of types which together form a circular
 * reference. The array of cyclesets is ordered such that no type reachable from
 * a given cycleset is reachable from any cycleset earlier in the array. In
 * other words, if you emit the cyclesets in order there will be no forward
 * references.
 */
export function createCycleSets(types: Type[]): Type[][] {
  const inputTypes = new Set(types);

  let index = 0; // Unique index assigned to each node

  const stack: Type[] = [];
  const onStack = new Set<Type>();

  /* Map of type to the type's index */
  const indices = new Map<Type, number>();

  /* Map to store the smallest type index reachable from the given type */
  const lowlink = new Map<Type, number>();

  const sccs: Type[][] = [];

  for (const node of types) {
    if (!indices.has(node)) {
      strongConnect(node);
    }
  }

  return sccs;

  function referencedTypes(type: Type): Type[] {
    switch (type.kind) {
      case "Model":
        return [
          ...(type.baseModel ? [type.baseModel] : []),
          ...(type.indexer ? [type.indexer.key, type.indexer.value] : []),
          ...[...type.properties.values()].map((p) => p.type),
        ];

      case "Union":
        return [...type.variants.values()].map((v) =>
          v.kind === "UnionVariant" ? v.type : v,
        );
      case "UnionVariant":
        return [type.type];
      case "Interface":
        return [...type.operations.values()];
      case "Operation":
        return [type.parameters, type.returnType];
      case "Enum":
        return [];
      case "Scalar":
        return type.baseScalar ? [type.baseScalar] : [];
      case "Tuple":
        return type.values;
      case "Namespace":
        return [
          ...type.operations.values(),
          ...type.scalars.values(),
          ...type.models.values(),
          ...type.enums.values(),
          ...type.interfaces.values(),
          ...type.namespaces.values(),
        ];
      default:
        return [];
    }
  }

  // The main recursive function that implements Tarjan's algorithm.
  function strongConnect(v: Type): void {
    // Set the depth index for v to the smallest unused index
    indices.set(v, index);
    lowlink.set(v, index);
    index++;
    stack.push(v);
    onStack.add(v);

    // Consider successors of v
    for (const w of referencedTypes(v)) {
      if (!indices.has(w)) {
        // Successor w has not yet been visited; recurse on it.
        strongConnect(w);
        // After recursion, update lowlink[v]
        lowlink.set(v, Math.min(lowlink.get(v)!, lowlink.get(w)!));
      } else if (onStack.has(w)) {
        // If w is in the current SCC (i.e. on the stack), update lowlink[v]
        lowlink.set(v, Math.min(lowlink.get(v)!, indices.get(w)!));
      }
    }

    // If v is a root node, pop the stack and generate an SCC.
    if (lowlink.get(v) === indices.get(v)) {
      const component: Type[] = [];
      let w: Type;
      do {
        w = stack.pop()!;
        onStack.delete(w);
        component.push(w);
      } while (w !== v);

      const scc = component.filter((v) => inputTypes.has(v));
      if (scc.length > 0) {
        sccs.push(scc);
      }
    }
  }
}

export function call(target: string, ...args: Children[]) {
  return <FunctionCallExpression target={target} args={args} />;
}
