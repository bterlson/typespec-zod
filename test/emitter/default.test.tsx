import { expect, it } from "vitest";
import { createEmitterTestRunner } from "../utils.jsx";

it("emits all declarations", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`
    model MyModel {
      id: string;
    }

    model MyModelArray is Array<MyModel> {}
    
    scalar MyScalar extends string;

    enum MyEnum {
      Value: 1
    }

    union MyFoo {
      1, 2
    }
  `);

  const { text } = await runner.program.host.readFile("typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});

it("handles references by doing a topological sort", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`
    union MyUnion {
      one: MyModelArray;
      two: MyModel;
    }

    model MyModelArray is Array<MyModel> {}

    model MyModel {
      id: string;
    }
  `);

  const { text } = await runner.program.host.readFile("typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});

it("handles the readme sample", async () => {
  const runner = await createEmitterTestRunner();
  await runner.compile(`

    model PetBase {
      age: uint8;
    
      @maxLength(20)
      name: string
    }
    
    model Dog extends PetBase {
      walksPerDay: safeint;
    }
    
    model Cat extends PetBase {
      belongingsShredded: uint64;
    }
    
    @discriminated
    union Pet {
      dog: Dog,
      cat: Cat,
    }
  `);

  const { text } = await runner.program.host.readFile("typespec-zod/models.ts");

  expect(text).toMatchSnapshot();
});
