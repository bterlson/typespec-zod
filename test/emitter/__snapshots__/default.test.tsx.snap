// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles the readme sample 1`] = `
"import { z } from "zod";

export const petBase = z.object({
  age: z.number().int().nonnegative().lte(255),
  name: z.string().max(10),
});

export const dog = z.petBase.merge(z.object({
  walksPerDay: z.number().int().safe(),
}));

export const cat = z.petBase.merge(z.object({
  belongingsShredded: z.bigint().nonnegative().lte(18446744073709551615),
}));

export const pet = z.discriminatedUnion(
  "kind",
  [
    z.object({
      kind: z.literal("dog"),
      value: dog,
    }),
    z.object({
      kind: z.literal("cat"),
      value: cat,
    })
  ]
);"
`;
